File: 014/Ruby/problem014.rb
$LOAD_PATH << "../../lib/Ruby"

class Integer
  
  @@collatz = Array.new(10000000)

  def collatz
    if self > @@collatz.length
      self.raw_collatz
    else
      @@collatz[self] ||= self.raw_collatz
    end      
  end

  def raw_collatz
    if self == 1
      return 1
    elsif self.even?
      (self / 2).collatz + 1
    else
      (3 * self + 1).collatz + 1
    end
  end

  def maximum_collatz_under
    index = -1
    max = -1
    (1..self).each do |i|
      collatzOfI = i.collatz
      if collatzOfI > max
        index, max = i, collatzOfI
      end
    end
    index
  end

end

File: 014/Python/problem014.py
def collatz(n):
    if n == 1:
        return 1
    elif n % 2 == 0:
        return collatz(n / 2) + 1
    else:
        return collatz(3 * n + 1) + 1
    
def maximum_collatz_under(n):
    if n == 1:
        return 1
    maximum = 0
    maxIndex = 0
    for i in range(1, n+1):
        collatzOfI = collatz(i)
        if collatzOfI > maximum:
            maximum = collatzOfI
            maxIndex = i
    return maxIndex    
